tests/tests.cpp

```diff
@@ -0,0 +1,398 @@
+            std::vector<uint8_t> perturbed_sig(sig);
+            perturb(perturbed_sig);
+            ASSERT_NE(0, crypto.Verify(pem_key, msg.data(),
+                                       perturbed_sig.data(), msg.size()))
+                << "Perturbed signature verified successfully";
+        }
+
+        // 2) Make random pertubations in the message
+        for (int i = 0; i < 100; i++) {
+            std::vector<uint8_t> perturbed_msg(msg);
+            perturb(perturbed_msg);
+            ASSERT_NE(0, crypto.Verify(pem_key, perturbed_msg.data(),
+                                       sig.data(), perturbed_msg.size()))
+                << "Perturbed message verified successfully";
+        }
+
+        // 3) Verify with the wrong key
+        // Load public key from file
+        std::vector<uint8_t> new_key;
+        load_file(std::string("../../tests/data/public.pub"), new_key);
+
+        ASSERT_NE(0, crypto.Verify(new_key.data(), msg.data(), sig.data(),
+                                   msg.size()))
+            << "Message verified with the incorrect key";
+    }
+
+    TEST_F(SoundnessTest, Attestation) {
+        // Currently the unittests only support building for the host, so we
+        // can't generate evidence. Instead we modify pre-generated evidence
+
+        // Load nonce, public key, enclave signing key, and evidence from file
+        std::vector<uint8_t> nonce;
+        std::vector<uint8_t> pub_key;
+        std::vector<uint8_t> signing_key;
+        std::vector<uint8_t> evidence;
+        load_file(std::string("../../tests/data/nonce"), nonce);
+        load_file(std::string("../../tests/data/report_key.pub"), pub_key);
+        load_file(std::string("../../tests/data/signing_key.pub"), signing_key);
+        load_file(std::string("../../tests/data/evidence"), evidence);
+
+        // Make random pertubations in the evidence
+        for (int i = 0; i < 100; i++) {
+            std::vector<uint8_t> perturbed_evidence(evidence);
+            perturb(perturbed_evidence);
+            ASSERT_NE(0, attestation->AttestEvidence(
+                             &sgx_remote_uuid, signing_key.data(),
+                             perturbed_evidence.data(), pub_key.data(),
+                             nonce.data(), signing_key.size() + 1,
+                             perturbed_evidence.size(), pub_key.size()))
+                << "Perturbed evidence verified successfully";
+        }
+
+        // Use a different nonce
+        uint8_t new_nonce[CIPHER_IV_SIZE];
+        crypto.RandGen(new_nonce, CIPHER_IV_SIZE);
+
+        ASSERT_NE(0, attestation->AttestEvidence(
+                         &sgx_remote_uuid, signing_key.data(), evidence.data(),
+                         pub_key.data(), new_nonce, signing_key.size() + 1,
+                         evidence.size(), pub_key.size()))
+            << "Evidence verified successfully with invalid nonce";
+
+        // Use an incorrect public key
```

Comment:
can we add a note saying that the correct key is `pub_key`, but we pass in `pem_key` a randomly generated key, so we expect it to fail

Reply:
Added